README
NAME

    Crypt::PKCS10 - parse PKCS #10 certificate requests

VERSION

    version 1.800201

SYNOPSIS

        use Crypt::PKCS10;
    
        Crypt::PKCS10->setAPIversion( 1 );
        my $decoded = Crypt::PKCS10->new( $csr ) or die Crypt::PKCS10->error;
    
        print $decoded;
    
        @names = $decoded->extensionValue('subjectAltName' );
        @names = $decoded->subject unless( @names );
    
        %extensions = map { $_ => $decoded->extensionValue( $_ ) } $decoded->extensions

DESCRIPTION

    Crypt::PKCS10 parses PKCS #10 certificate requests (CSRs) and provides
    accessor methods to extract the data in usable form.

    Common object identifiers will be translated to their corresponding
    names. Additionally, accessor methods allow extraction of single data
    fields. The format of returned data varies by accessor.

    The access methods return the value corresponding to their name. If
    called in scalar context, they return the first value (or an empty
    string). If called in array context, they return all values.

    true values should be specified as 1 and false values as 0. Future API
    changes may provide different functions when other values are used.

      This file is automatically generated by pod2readme from PKCS10.pm and Changes.

NAME

    Crypt::PKCS10 - parse PKCS #10 certificate requests

RELEASE NOTES

    Version 1.4 made several API changes. Most users should have a painless
    migration.

    ALL users must call Crypt::PKCS10->setAPIversion. If not, a warning
    will be generated by the first class method called. This warning will
    be made a fatal exception in a future release.

    Other than that requirement, the legacy mode is compatible with
    previous versions.

    new will no longer generate exceptions. undef is returned on all
    errors. Use the error class method to retrieve the reason.

    new will accept an open file handle in addition to a request.

    Users are encouraged to migrate to the version 1 API. It is much easier
    to use, and does not require the application to navigate internal data
    structures.

    Version 1.7 provides support for DSA and ECC public keys. By default,
    it verifies the signature of CSRs. It also allows the caller to verify
    the signature of a CSR. subjectPublicKeyParams and signatureParams
    provide additional information. The readFile option to new() will
    open() a file containing a CSR by name. The ignoreNonBase64 option
    allows PEM to contain extraneous characters. Changes describes
    additional improvements. Details follow.

INSTALLATION

    Crypt::PKCS10 supports DSA, RSA and ECC public keys in CSRs.

    It depends on Crypt::OpenSSL::DSA, Crypt::OpenSSL::RSA and
    Crypt::PK::ECC for some operations. All are recommended. Some methods
    will return errors if Crypt::PKCS10 is presented with a CSR containing
    an unsupported public key type.

    To install this module type the following:

        perl Makefile.PL
        make
        make test
        make install

REQUIRES

    Convert::ASN1

    Crypt::OpenSSL::DSA

    Crypt::OpenSSL::RSA

    Crypt::PK::ECC

    Digest::SHA

    For ECC: Crypt::PK::ECC

    Very old CSRs may require DIGEST::MD{5,4,2}

METHODS

    Access methods may exist for subject name components that are not
    listed here. To test for these, use code of the form:

      $locality = $decoded->localityName if( $decoded->can('localityName') );

    If a name component exists in a CSR, the method will be present. The
    converse is not (always) true.

 class method setAPIversion( $version )

    Selects the API version (0 or 1) expected.

    Must be called before calling any other method.

    The API version determines how a CSR is parsed. Changing the API
    version after parsing a CSR will cause accessors to produce
    unpredictable results.

    Version 0 - DEPRECATED

      Some OID names have spaces and descriptions

      This is the format used for Crypt::PKCS10 version 1.3 and lower. The
      attributes method returns legacy data.

    Version 1

      OID names from RFCs - or at least compatible with OpenSSL and ASN.1
      notation. The attributes method conforms to version 1.

    If not called, a warning will be generated and the API will default to
    version 0.

    In a future release, the warning will be changed to a fatal exception.

    To ease migration, both old and new names are accepted by the API.

    Every program should call setAPIversion(1).

 class method getAPIversion

    Returns the current API version.

    Returns undef if setAPIversion has never been called.

 class method new( $csr, %options )

    Constructor, creates a new object containing the parsed PKCS #10
    certificate request.

    $csr may be a scalar containing the request, a file name, or a file
    handle from which to read it.

    If a file name is specified, the readFile option must be specified.

    If a file handle is supplied, the caller should specify acceptPEM => 0
    if the contents are DER.

    The request may be PEM or binary DER encoded. Only one request is
    processed.

    If PEM, other data (such as mail headers) may precede or follow the
    CSR.

        my $decoded = Crypt::PKCS10->new( $csr ) or die Crypt::PKCS10->error;

    Returns undef if there is an I/O error or the request can not be parsed
    successfully.

    Call error() to obtain more detail.

  options

    The options are specified as name => value.

    If the first option is a HASHREF, it is expanded and any remaining
    options are added.

    acceptPEM

      If false, the input must be in DER format. binmode will be called on
      a file handle.

      If true, the input is checked for a CERTIFICATE REQUEST header. If
      not found, the csr is assumed to be in DER format.

      Default is true.

    PEMonly

      If true, the input must be in PEM format. An error will be returned
      if the input doesn't contain a CERTIFICATE REQUEST header. If false,
      the input is parsed according to acceptPEM.

      Default is false.

    binaryMode

      If true, an input file or file handle will be set to binary mode
      prior to reading.

      If false, an input file or file handle's binmode will not be
      modified.

      Defaults to false if acceptPEM is true, otherwise true.

    escapeStrings

      If true, strings returned for extension and attribute values are
      '\'-escaped when formatted. This is compatible with OpenSSL
      configuration files.

      The special characters are: '\', '$', and '"'

      If false, these strings are not '\'-escaped. This is useful when they
      are being displayed to a human.

      The default is true.

    ignoreNonBase64

      If true, most invalid base64 characters in PEM data will be ignored.
      For example, this will accept CSRs prefixed with '> ', as e-mail when
      the PEM is inadvertently quoted. Note that the BEGIN and END lines
      may not be corrupted.

      If false, invalid base64 characters in PEM data will cause the CSR to
      be rejected.

      The default is false.

    readFile

      If true, $csr is the name of a file containing the CSR.

      If false, $csr contains the CSR or is an open file handle.

      The default is false.

    verifySignature

      If true, the CSR's signature is checked. If verification fails, new
      will fail.

      If false, the CSR's signature is not checked.

      The default is true for API version 1 and false for API version 0.

    No exceptions are generated.

    The defaults will accept either PEM or DER from a string or file hande,
    which will not be set to binary mode. Automatic detection of the data
    format may not be reliable on file systems that represent text and
    binary files differently. Set acceptPEM to false and PEMonly to match
    the file type on these systems.

    The object will stringify to a human-readable representation of the
    CSR. This is useful for debugging and perhaps for displaying a request.
    However, the format is not part of the API and may change. It should
    not be parsed by automated tools.

    Exception: The public key and extracted request are PEM blocks, which
    other tools can extract.

    If another object inherits from Crypt::PKCS10, it can extend the
    representation by overloading or calling as_string.

 {class} method error

    Returns a string describing the last error encountered;

    If called as an instance method, last error encountered by the object.

    If called as a class method, last error encountered by the class.

    Any method can reset the string to undef, so the results are only valid
    immediately after a method call.

 class method name2oid( $oid )

    Returns the OID corresponding to a name returned by an access method.

    Not in API v0;

 csrRequest( $format )

    Returns the binary (ASN.1) request (after conversion from PEM and
    removal of any data beyond the length of the ASN.1 structure.

    If $format is true, the request is returned as a PEM CSR. Otherwise as
    a binary string.

 certificationRequest

    Returns the binary (ASN.1) section of the request that is signed by the
    requestor.

    The caller can verify the signature using signatureAlgorithm,
    certificationRequest and signature(1).

 Access methods for the subject's distinguished name

    Note that subjectAltName is prefered, and that modern certificate users
    will ignore the subject if subjectAltName is present.

  subject( $format )

    Returns the entire subject of the CSR.

    In scalar context, returns the subject as a string in the form
    /componentName=value,value. If format is true, long component names are
    used. By default, abbreviations are used when they exist.

      e.g. /countryName=AU/organizationalUnitName=Big org/organizationalUnitName=Smaller org
      or     /C=AU/OU=Big org/OU=Smaller org

    In array context, returns an array of (componentName, [values]) pairs.
    Abbreviations are not used.

    Note that the order of components in a name is significant.

  commonName

    Returns the common name(s) from the subject.

        my $cn = $decoded->commonName();

  organizationalUnitName

    Returns the organizational unit name(s) from the subject

  organizationName

    Returns the organization name(s) from the subject.

  emailAddress

    Returns the email address from the subject.

  stateOrProvinceName

    Returns the state or province name(s) from the subject.

  countryName

    Returns the country name(s) from the subject.

 subjectAltName( $type )

    Convenience method.

    When $type is specified: returns the subject alternate name values of
    the specified type in list context, or the first value of the specified
    type in scalar context.

    Returns undefined/empty list if no values of the specified type are
    present, or if the subjectAltName extension is not present.

    Types can be any of:

       otherName
     * rfc822Name
     * dNSName
       x400Address
       directoryName
       ediPartyName
     * uniformResourceIdentifier
     * iPAddress
     * registeredID

    The types marked with '*' are the most common.

    If $type is not specified: In list context returns the types present in
    the subjectAlternate name. In scalar context, returns the SAN as a
    string.

 version

    Returns the structure version as a string, e.g. "v1" "v2", or "v3"

 pkAlgorithm

    Returns the public key algorithm according to its object identifier.

 subjectPublicKey( $format )

    If $format is true, the public key will be returned in PEM format.

    Otherwise, the public key will be returned in its hexadecimal
    representation

 subjectPublicKeyParams

    Returns a hash describing the public key. The contents vary depending
    on the public key type.

  Standard items:

    keytype - ECC, RSA, DSA or undef

    keytype will be undef if the key type is not supported. In this case,
    error() returns a diagnostic message.

    keylen - Approximate length of the key in bits.

    Other items include:

    For RSA, modulus and publicExponent.

    For DSA, G, P and Q.

    For ECC, curve, pub_x and pub_y. curve is an OID name.

  Additional detail

    subjectPublicKeyParams(1) returns the standard items, and may also
    return detail, which is a hashref.

    For ECC, the detail hash includes the curve definition constants.

 signatureAlgorithm

    Returns the signature algorithm according to its object identifier.

 signatureParams

    Returns the parameters associated with the signatureAlgorithm as
    binary. Returns undef if none, or if NULL.

    Note: In the future, some signatureAlgorithms may return a hashref of
    decoded fields.

    Callers are advised to check for a ref before decoding...

 signature( $format )

    The CSR's signature is returned.

    If $format is 1, in binary.

    If $format is 2, decoded as an ECDSA signature - returns hashref to r
    and s.

    Otherwise, in its hexadecimal representation.

 attributes( $name )

    A request may contain a set of attributes. The attributes are OIDs with
    values. The most common is a list of requested extensions, but other
    OIDs can also occur. Of those, challengePassword is typical.

    For API version 0, this method returns a hash consisting of all
    attributes in an internal format. This usage is deprecated.

    For API version 1:

    If $name is not specified, a list of attribute names is returned. The
    list does not include the requestedExtensions attribute. For that, use
    extensions();

    If no attributes are present, the empty list (undef in scalar context)
    is returned.

    If $name is specified, the value of the extension is returned. $name
    can be specified as a numeric OID.

    In scalar context, a single string is returned, which may include lists
    and labels.

      cspName="Microsoft Strong Cryptographic Provider",keySpec=2,signature=("",0)

    Special characters are escaped as described in options.

    In array context, the value(s) are returned as a list of items, which
    may be references.

     print( " $_: ", scalar $decoded->attributes($_), "\n" )
                                       foreach ($decoded->attributes);

    See the module documentation for a list of known OID names.

    It is too long to include here.

 extensions

    Returns an array containing the names of all extensions present in the
    CSR. If no extensions are present, the empty list is returned.

    The names vary depending on the API version; however, the returned
    names are acceptable to extensionValue, extensionPresent, and name2oid.

    The values of extensions vary, however the following code fragment will
    dump most extensions and their value(s).

     print( "$_: ", $decoded->extensionValue($_,1), "\n" ) foreach ($decoded->extensions);

    The sample code fragment is not guaranteed to handle all cases.
    Production code needs to select the extensions that it understands and
    should respect the critical boolean. critical can be obtained with
    extensionPresent.

 extensionValue( $name, $format )

    Returns the value of an extension by name, e.g. extensionValue(
    'keyUsage' ). The name SHOULD be an API v1 name, but API v0 names are
    accepted for compatibility. The name can also be specified as a numeric
    OID.

    If $format is 1, the value is a formatted string, which may include
    lists and labels. Special characters are escaped as described in
    options;

    If $format is 0 or not defined, a string, or an array reference may be
    returned. The array many contain any Perl variable type.

    To interpret the value(s), you need to know the structure of the OID.

    See the module documentation for a list of known OID names.

    It is too long to include here.

 extensionPresent( $name )

    Returns true if a named extension is present: If the extension is
    critical, returns 2. Otherwise, returns 1, indicating not critical, but
    present.

    If the extension is not present, returns undef.

    The name can also be specified as a numeric OID.

    See the module documentation for a list of known OID names.

    It is too long to include here.

 registerOID( )

    Class method.

    Register a custom OID, or a public OID that has not been added to
    Crypt::PKCS10 yet.

    The OID may be an extension identifier or an RDN component.

    The oid is specified as a string in numeric form, e.g. '1.2.3.4'

  registerOID( $oid )

    Returns true if the specified OID is registered, false otherwise.

  registerOID( $oid, $longname, $shortname )

    Registers the specified OID with the associated long name. This enables
    the OID to be translated to a name in output.

    The long name should be Hungarian case (commonName), but this is not
    currently enforced.

    Optionally, specify the short name used for extracting the subject. The
    short name should be upper-case (and will be upcased).

    E.g. built-in are $oid => '2.4.5.3', $longname => 'commonName',
    $shortname => 'CN'


Changes
Revision history for perl module Crypt::PKCS10

1.0 2014-08-20

 - Initial Version

1.1 2015-10-09

 - Allow same OID elements in RDNs
 - Support for all DirectoryStrings
 - New accessor method: organizationName 

1.2 2015-11-13

 - Add generic method extensionValue
 - Ommit unknown or damaged extensions

1.3 2015-11-17

 - Experimental Perl features removed
 - Refactored building process

1.4_01 2016-01-14

- Find PEM anywhere
- Support repeated attributes
- Various crashes, internal optimizations
- More OIDs
- Access methods: subject, subjectAltName, extensions
- Generate access methods for unknown OIDs in DNs at runtime
- Add extractCSR method to get CSR in binary or PEM
- subjectPublicKey option to extract in PEM format
- Convert IP addresses to strings
- Convert basicConstraints to string
- Add registerOID to allow extraction of simple custom OIDs
- More tests
- Improve documentation
- Add setAPIversion, regularize OID names in V1
- For API V1, attributes method returns names or values of attributes (except extensions)
- Improve build: Use Dist::Zilla, autogenerate README,LICENE,META,Makefile.PL
- Generate & ship Commitlog from git

1.4_02

- For API V1, omit space in key usage list
- Look for policy identifier names in all tables
- Decode certificatePolicyIdentifier, add related OIDs
- Avoid building un-necessary ASN.1 parsers
- Fully decode ApplicationCertPolicies for API V1
- Check enhanced key usage before registering OID
- Add name2oid to API
- Internal cleanup  & error reporting improvements

1.4_03

- certificateTemplate ASN correction: MajorVersion is optional
- certificateTemplate returns id as a name if OID registered (API V1)
- handle some malformed requests better
- decode more M$ attributes, correct some.
- Don't croak() on missing attributes
- Convert BMPStrings from UCS2 to Perl (printable) strings
- Rework special OID handling to be scalable & cover attributes
- Don't croak in new.  (API V1) Always return undef.  Class method error() returns detail.
- Return reasonable strings from attributes('name') and extensionValue('name',1)
- Overload object so print can produce a useful dump.
- Handle null subject
- More enhanced key usage OIDs
- keyUsage returns array rather than comma separated string (API v1)
- Cleaned up POD
- Do a better job with error reporting
- Validate PEM's Base64 encoding.
- Ensure that only the first PEM certificate is extracted.
- Improve tests
- Enable (internal) custom formatters for extensions and attributes.
- Return basicConstraints as a hash
- Accept numeric OIDs as arguments to extensionValue, extensionPresent and attributes.

1.4_04
 - Build Kwalitee issues: Minimum Perl version, Repo location, and Provides in META
 - Fix test issues due to hash randomization on newer Perl versions.

1.4_05
 - Sign distribution kits
 - Remove execute permissions from text files
 - Generate Markdown format README for github
 - Use Pod::Weaver to manage authorship, copyright from dist.ini
 - Add prerequisites display test
 - Validate with perlcritic, pod syntax

1.5
 - Uncoordinated release of 1.4_04

1.6
 - Support some EC OIDs

1.7
 - ECC test doesn't require Data::Dumper
 - Support more EC OIDs
 - Add DSA OIDs
 - Add subjectPublicKeyParams(), which provides some description of the key
 - Add certificationRequest and signature(1) to enable verification of CSR signatures
 - Add signatureParams()
 - Add signature(2) to extract ECDSA signature components.
 - Fix and extend registerOID, add tests.
 - Add examples/ directory with sample code
 - Verify signature of CSRs in new() by default.  Add checkSignature().
 - Optionally allow invalid base64 character in PEM data.
 - Allow new() from filename and options from hashref.

1.7_01
 - Address CPANTS failures due to bugs in old versions of Crypt::PK::ECC
 - Add explicit version dependency for CryptX, the parent distribution.
 - For insurance, add explicit version requirements for the other modules we depend on.
 - Add getAPIversion
 - Make stringify use an overloadable method so it can be extended if subclassed.
 - Correctly parse DER for expected length when removing padding.

1.8
 - When returning PEM, ensure max line width of 64 to satisfy RFC 7468, not 76 of MIME.
 - Fix test failures on some editions of 5.8.8 due to Perl bug 39185.  (Tripped in test, not module.)

1.8001
 - Makefile.pl: Diagnose old OpenSSL versions that fail signature verification and discard automated tests in that case.
 - Print full OpenSSL configuration during automated testing.
 - Require CryptX with version fix.
 - Silence test warning if openssl command is missing.
 - No changes to functional code.

1.8002
 - Cosmetic: wrap long form OpenSSL config in test
 - Make RSA, DSA and ECC support optional.  Warn during install none are present or any don't work.
 - Signature verification requires the corresponding module, as does decoding ECC public keys.
 - subjectPublicKeyParams and checkSignature set error() for unsupported key type or hash.
 - Clean up carp/croak, especially where crept into evals.
 - Provide instance method for error.
 - Fix stringify's interactions with API v0.
 - Add binaryMode and PEMonly options to provide finer grained control of formats

1.8002_01
 - Filter OpenSSL's 'WARNING: can't open config file' messages.
 - Add dieOnError option to new()
 - Report undefined $csr gracefully
 - API v0 tests should not try to verify a signature.
 - new() in void context will generate an exception.


__END__


Makefile.PL
# This Makefile.PL for Crypt-PKCS10 was generated by
# Dist::Zilla::Plugin::MakeMaker::Awesome 0.38.
# Don't edit it but the dist.ini and plugins used to construct it.

use strict;
use warnings;

use 5.008008;
use ExtUtils::MakeMaker;

# -*- mode: cperl; -*-
#
# This fragment is inserted into the Makefile.PL to ensure that signature verification works.
# It has been noted that old versions of OpenSSL will satisfy the Crypt::OpenSSL::xx modules,
# but not support verification.  The hex strings are CSRs and hash values that should verify.
# Hex is used for portability and readability.
#
# Here, if a support module is present, it is tested and must work.
# All modules are optional - but at least one should be present.
{
    my $avail = 0;

    print STDERR ( "Checking crypto modules' operation with OpenSSL library ..." );

    my( @nover, @failtext );

    my $ok = eval {
        require Crypt::OpenSSL::DSA;
    };
    if( $ok ) {
        print STDERR ( " DSA" );
        eval {
            my $key = Crypt::OpenSSL::DSA->read_pub_key_str( << '_KEY_' ) or die "Can't read DSA public key";
-----BEGIN PUBLIC KEY-----
MIIBtzCCASwGByqGSM44BAEwggEfAoGBAOs6x6eSjwoque9hKIz94RwT6TLThTgD
2uslWeipGrydxIV3GVpHECbvJ3QfJOYNk6QlBvFs2L1a69v1GbW6o+ZHBITDw3kP
/JtWF/vThUXNB/9g2nhGODyEjwq0R6x+1dzTUTLYguAyafNpQzDUEpLZLkRyQp/6
DiUU7DXqlu4tAhUAsqEwY1v+Gduz5J2PXEuugmYSYBkCgYEA0qgvsy8wOqt8VUyR
CW0jPNPoeyyeICFypSBseiKKORlVBPz2JmdI6hohLO9rljK9wgEqdmh1yTM099rM
JP727RHBha9QKyNmN7/bP4+rHeK0vCa0XVu2FxuMFp7Kd5d7W0ucnKffQFLHcXvY
hduUNtCYKWWeiG3jUXPaU6FreNcDgYQAAoGAFVWpRu48Az2VrgwuYohcT9u9HlEb
xj9g9cYD8Jb/Hg9uaH3z+dus1ocuqnVgsaGRzBvlHYOwx8PYrvkebeuryShmpgOp
Zcs55fpyS9n7ML9XYFlSWnE4p1jOlHM6Am1XwZkGQ/0wCVcWpFNLNnlEIEPaic7F
Gs0wfthYSf4FK+I=
-----END PUBLIC KEY-----
_KEY_
            return $key->verify( pack( 'H*', 'a1d525953bc995b5c2568d5797b0e7525929bb7a46db1187c62d4ea3a040b746' ),
                                 pack( 'H*', '302c021439b795be2af338ae34384fa8b235b67c2bbdeeb1021471ce864c3c1a' .
                                             'f74a8eb04a32dfeea17cc0ad2ac7' ) );
        };
        if( $@ || !$ok ) {
            push @nover, 'DSA';
            chomp $@;
            push @failtext, qq{Failed to verify DSA signature: $@};
            print STDERR ( '-' );
        } else {
            ++$avail;
            print STDERR ( '+' );
        }
    } else {
        print STDERR ( " !DSA" );
    }

    $ok = eval {
        require Crypt::OpenSSL::RSA;
    };
    if( $ok ) {
        print STDERR ( " RSA" );
        eval {
            my $key = Crypt::OpenSSL::RSA->new_public_key( << '_KEY_' ) or die "Can't read RSA public key";
-----BEGIN PUBLIC KEY-----
MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA4EhMEu4ppW+3LSgp/fKG
hZsEmgB9kDASa90enSMZvji0pAsAQW3FSwADQLpYC7HFEeJR4aeB7CE5xS1B4WIm
9gfRxLMCekqVHq3IjpCxAN5WjyZ5AsaUOZ0TkrJ7en8x2EeV5R1oM+5GEyv8BJ+f
lizG9Q5RHxpWIn1H1+PWD4dW2RSo/PVECmflceQQb6bmyxy+bka5Sr7WLxG95LLP
ss8zBVhlTn8nzMgrKHCFF6MzajapMItWg8vz3MpJLNVjrjp00tM3QkpkR3HM6HBN
xH5n7P8jiVh6V+OiGXgTEUpYzs0mAHG/A8l6pLLQvw4fUTECArx97nm6nohKZSij
bwIDAQAB
-----END PUBLIC KEY-----
_KEY_
            $key->use_sha256_hash;
            $key->use_pkcs1_padding;

            return $key->verify( pack( 'H*', '308201b602010030818831133011060a0992268993f22c64011916036f'.
                                             '726731173015060a0992268993f22c64011916074f70656e53534c3115'.
                                             '3013060a0992268993f22c640119160575736572733123300b06035504'.
                                             '030c04746573743014060a0992268993f22c6401010c06313233343536'.
                                             '311c301a06092a864886f70d010901160d7465737440746573742e636f'.
                                             '6d30820122300d06092a864886f70d01010105000382010f003082010a'.
                                             '0282010100e0484c12ee29a56fb72d2829fdf286859b049a007d903012'.
                                             '6bdd1e9d2319be38b4a40b00416dc54b000340ba580bb1c511e251e1a7'.
                                             '81ec2139c52d41e16226f607d1c4b3027a4a951eadc88e90b100de568f'.
                                             '267902c694399d1392b27b7a7f31d84795e51d6833ee46132bfc049f9f'.
                                             '962cc6f50e511f1a56227d47d7e3d60f8756d914a8fcf5440a67e571e4'.
                                             '106fa6e6cb1cbe6e46b94abed62f11bde4b2cfb2cf330558654e7f27cc'.
                                             'c82b28708517a3336a36a9308b5683cbf3dcca492cd563ae3a74d2d337'.
                                             '424a644771cce8704dc47e67ecff2389587a57e3a2197813114a58cecd'.
                                             '260071bf03c97aa4b2d0bf0e1f51310202bc7dee79ba9e884a6528a36f'.
                                             '0203010001a000' ),
                                 pack( 'H*', 'dc8ba14eade00b952cdaacf0f48986407f01802b3d2c626c0cea42f977'.
                                             'a39c3dbb58db01c1897fb72a3d46c811269e446c4c208f63030694216f'.
                                             '84ecca99163a60587e5bdf14075728e49e9b5e3992672021e694c1dcc2'.
                                             'a5f5ba4fb2c02de1453702ffe93ec2d9dbb3486f0a1e9cb75ee24e6f6a'.
                                             '23c4e57f8c611c5aec5de08e0a6b4ddff1d28f2ddcf5d61507f8dee490'.
                                             'deb185a20f2d52b29e43fc583d3b00b7ee9e7b54d2c4c828f7f7babf6d'.
                                             'd42d77dc2431029b61330e0b55a6dcc728f573af196fdebecdb4b78c48'.
                                             '3636e4c537fe36f69d86c5a7e024672a4504caa3a2be48e891ff4b83b4'.
                                             '815ea6972f54367466decce6be94c67aed04145392684726' ) );
        };
        if( $@ || !$ok ) {
            push @nover, 'RSA';
            chomp $@;
            push @failtext, qq{Failed to verify RSA signature: $@};
            print STDERR ( '-' );
        } else {
            ++$avail;
            print STDERR ( '+' );
        }
    } else {
        print STDERR ( '!RSA' );
    }

    $ok = eval {
        require Crypt::PK::ECC;
    };
    if( $ok ) { # ECC installed, see if it works.
        print STDERR ( " ECC" );
        eval {
            my $key = Crypt::PK::ECC->new( \<< '_KEY_' );
-----BEGIN PUBLIC KEY-----
MFowFAYHKoZIzj0CAQYJKyQDAwIIAQEHA0IABI0FB6fr9YoXkQ/isVsMRR6TvJSK
S6+3vxIE1gQ+feE5QjC++rnFEVzTzR4FmlRXiLseCDDuBjAMTz6NhxKPPdw=
-----END PUBLIC KEY-----
_KEY_
            return $key->verify_message( pack( 'H*', '30440220730d25ebe5f187c607577cc106d3141dc7f9082791'.
                                                     '4f2a6a11ebc9de6fdf1d26022042c02e4819f2c16c56181205'.
                                                     'c6c2176902f20cbfcfdc1fa82b30f79bd15d2172' ),
                                         pack( 'H*', '3081a80201003045310b300906035504061302415531133011'.
                                                     '06035504080c0a536f6d652d53746174653121301f06035504'.
                                                     '0a0c18496e7465726e6574205769646769747320507479204c'.
                                                     '7464305a301406072a8648ce3d020106092b24030302080101'.
                                                     '07034200048d0507a7ebf58a17910fe2b15b0c451e93bc948a'.
                                                     '4bafb7bf1204d6043e7de1394230befab9c5115cd3cd1e059a'.
                                                     '545788bb1e0830ee06300c4f3e8d87128f3ddca000' ),
                                         'SHA256' );
        };
        if( $@ || !$ok ) {
            push @nover, 'ECC';
            chomp $@;
            push @failtext, qq{Failed to verify ECC signature: $@};
            print STDERR ( '-' );
        } else {
            ++$avail;
            print STDERR ( '+' );
        }
    } else {
        print STDERR ( " !ECC" );
    }

    if( @nover || $ENV{CRYPT_PKCS10_DUMPCFG} ) {
        # At least one signature failed to verify
        #
        # Do not allow Makefile creation.  Provide diagnostic.
        # Exit providing status for automated testing.

        my @removed;
        unlink( $_ ) && push @removed, $_ foreach( qw/Makefile Build/ );

        my $sslver = eval {
            local $SIG{__WARN__} = sub {};

            my $text = qx/openssl version -a 2>&1/;
            return unless( $? == 0 && defined $text && length $text &&
                           $text !~ /invalid command/ );

            # list-public-key-algorithms would be cleaner, but isn't present in
            # OpenSSL 0.9.8, which is still in use in 2016.

            if( 0 ) {
                my $algorithms = qx/openssl list-public-key-algorithms 2>&1/;
                if( $? == 0 && defined $algorithms && length $algorithms &&
                    $algorithms !~ /invalid command/ ) {
                    $algorithms = join( ' ',
                                        sort map { /^\s*OID:\s*(.*)$/? (qq("$1"),) : ()  }
                                        split( /\n/, $algorithms ) );
                    $text .= sprintf( "algorithms:   %s\n", $algorithms );
                }
            } else {
                my $ciphers = qx/openssl ciphers 2>&1/;
                if( $? == 0 && defined $ciphers && length $ciphers &&
                    $ciphers !~ /invalid command/ ) {
                    chomp $ciphers;
                    $ciphers = join( ' ', sort split( /:/, $ciphers ) );
                    $text .= sprintf( "ciphers:      %s\n", $ciphers );
                }
            }

            require Text::Wrap;

            $text =~ s/^((?:compiler|options|ciphers|algorithms):[ ]+)([^\n]*)\n/
                       $1 . Text::Wrap::wrap( "", ' ' x 20, $2 ) . "\n"/gmsexi;

            return $text;
        };
        if( defined $sslver && length $sslver ) {
            chomp $sslver;
            $sslver = sprintf( "    Installed version of OpenSSL:\n%s", $sslver );
            $sslver =~ s/\n/\n      /g;
            $sslver .= "\n";
        } else {
            $sslver = '';
        }

        # Include the following to make the test result 'na', omit for 'discard'
        #     OS unsupported
        # 'na' will produce reports not counted as FAIL; 'discard' won't report at all.

        my $diag = sprintf( qq(

    Unable to verify %s signatures.  You probably need to upgrade OpenSSL.
    You may also need to install or rebuild Crypt::OpenSSL::DSA, Crypt::PK::ECC,
    and/or Crypt::OpenSSL::RSA.  These support respectively DSA, ECC and RSA
    public keys in CSRs.  At least one must be installed.

    Note: These modules may install successfully even if OpenSSL does not support
    the corresponding encryption method.  However, they will not work.  In this
    case, you must either upgrade OpenSSL to include the encryption method or
    uninstall the Perl module.

    No Makefile will be generated for Crypt::PKCS10.

    %s\n%s), (@nover? join( ', ', @nover ): 'any'), join( "\n    ", @failtext ), $sslver );

        $diag .=  sprintf( "\nDeleted old %s\n", join( ', ', @removed ) ) if( @removed );

        warn( $diag );
        exit 0;
    }
    if( $avail ) {
        print STDERR ( " ... OK\n" );
    } else {
        warn( qq{

    No signatures can be verified because none of the recommended modules is installed.

    Please install Crypt::OpenSSL::DSA, Crypt::PK::ECC, and/or Crypt::OpenSSL::RSA.
}           );
    }
}
# ####### End of SSL Test

my %WriteMakefileArgs = (
  "ABSTRACT" => "parse PKCS #10 certificate requests",
  "AUTHOR" => "Gideon Knocke <gknocke\@cpan.org>, Timothe Litt <tlhackque\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "Text::Wrap" => "2005.082401"
  },
  "DISTNAME" => "Crypt-PKCS10",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.008008",
  "NAME" => "Crypt::PKCS10",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Convert::ASN1" => "0.27",
    "Digest::MD2" => "2.03",
    "Digest::MD4" => "1.5",
    "Digest::MD5" => "2.51",
    "Digest::SHA" => "5.95",
    "Encode" => 0,
    "MIME::Base64" => 0,
    "Scalar::Util" => 0,
    "bytes" => 0,
    "overload" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Module::Signature" => 0,
    "Socket" => 0,
    "Test::More" => "0.94",
    "Text::Wrap" => "2005.082401"
  },
  "VERSION" => "1.8002",
  "test" => {
    "TESTS" => "t/*.t"
  }
);

my %FallbackPrereqs = (
  "Carp" => 0,
  "Convert::ASN1" => "0.27",
  "Digest::MD2" => "2.03",
  "Digest::MD4" => "1.5",
  "Digest::MD5" => "2.51",
  "Digest::SHA" => "5.95",
  "Encode" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "MIME::Base64" => 0,
  "Module::Signature" => 0,
  "Scalar::Util" => 0,
  "Socket" => 0,
  "Test::More" => "0.94",
  "Text::Wrap" => "2005.082401",
  "bytes" => 0,
  "overload" => 0,
  "strict" => 0,
  "warnings" => 0
);

unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
